### PR TITLE
CAMEL-9399: Added support for RabbitMQ Publisher Acknowledgements

### PR DESCRIPTION
**Pull request by FingolfinTEK**
Added support for RabbitMQ Publisher Acknowledgements, did a little code cleanup, added two producer integration tests that test out publisher acknowledgements.

### CODE REVIEW

File: components/camel-rabbitmq/src/main/java/org/apache/camel/component/rabbitmq/RabbitMQConnectionFactorySupport.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
+package org.apache.camel.component.rabbitmq;
+
+import java.security.KeyManagementException;
+import java.security.NoSuchAlgorithmException;
+
+import com.rabbitmq.client.ConnectionFactory;
+
+/**
+ * @author Milos Milivojevic | mmilivojevic@deployinc.com
```
Comment (davsclaus) [inline components/camel-rabbitmq/src/main/java/org/apache/camel/component/rabbitmq/RabbitMQConnectionFactorySupport.java]: We do not include author tags in our javadoc at Apache Camel. Do you mind removing that?

File: components/camel-rabbitmq/src/main/java/org/apache/camel/component/rabbitmq/RabbitMQEndpoint.java [hunk 2]
```diff
@@ -34,26 +26,11 @@
...
-
 import org.apache.camel.Consumer;
-import org.apache.camel.Exchange;
-import org.apache.camel.Message;
-import org.apache.camel.NoTypeConversionAvailableException;
-import org.apache.camel.Processor;
-import org.apache.camel.Producer;
-import org.apache.camel.RuntimeCamelException;
-import org.apache.camel.TypeConversionException;
+import org.apache.camel.*;
```
Comment (davsclaus) [inline components/camel-rabbitmq/src/main/java/org/apache/camel/component/rabbitmq/RabbitMQEndpoint.java]: please do not do wildcard imports. See building with checkstyle here to make the code comply: http://camel.apache.org/building

File: components/camel-rabbitmq/src/main/java/org/apache/camel/component/rabbitmq/RabbitMQEndpoint.java [hunk 3]
```diff
@@ -159,6 +136,10 @@
     private long requestTimeoutCheckerInterval = 1000;
     @UriParam
     private boolean transferException;
+    @UriParam(label = "producer", defaultValue = "false")
```
Comment (davsclaus) [inline components/camel-rabbitmq/src/main/java/org/apache/camel/component/rabbitmq/RabbitMQEndpoint.java]: false is implied as default, so no need.

File: components/camel-rabbitmq/src/main/java/org/apache/camel/component/rabbitmq/RabbitMQEndpoint.java [hunk 4]
```diff
@@ -159,6 +136,10 @@
...
+    private boolean publisherAcknowledgements;
+    @UriParam(label = "producer", defaultValue = "0")
```
Comment (davsclaus) [inline components/camel-rabbitmq/src/main/java/org/apache/camel/component/rabbitmq/RabbitMQEndpoint.java]: The same is 0 for a primitive long value

File: components/camel-rabbitmq/src/main/java/org/apache/camel/component/rabbitmq/RabbitMQMessageConverter.java [hunk 5]
```diff
@@ -217,4 +222,63 @@ private Object getValidRabbitMQHeaderValue(Object headerValue) {
...
+            }
+        }
+
+        if (hasSerializeHeader(properties)) {
+            Object messageBody = null;
+            try (InputStream b = new ByteArrayInputStream(body);
+                 ObjectInputStream o = new ObjectInputStream(b);) {
+                messageBody = o.readObject();
+            } catch (IOException | ClassNotFoundException e) {
+                LOG.warn("Could not deserialize the object");
```
Comment (davsclaus) [inline components/camel-rabbitmq/src/main/java/org/apache/camel/component/rabbitmq/RabbitMQMessageConverter.java]: We should not swallow exceptions. Maybe better to set the exception on the exchange or something ?

**Comment by davsclaus** [comment]
As you do both refactor the code and move stuff around etc. This would be better with 2 PRs - one about the refactor. And another PR about only the new code that is needed for this new functionality. Its hard to see what you do when its all together.

**Comment by FingolfinTEK** [comment]
Hi Claus, I've split the PRs as per your request; here you will find the changes for the additional logic. Once you review and merge it I will create another one for the refactoring (which consisted of moving some conversion logic out of the endpoint class and into the existing RabbitMQMessageConverter utility and of extracting connection factory creation and message publishing logic into separate classes to reduce the complexity of the endpoint class).

Thanks!

**Pull request closed without merging**